### DISCUSSION TITLE
Convergence and Local/Global Minima for PS problem

### DISCUSSION THREAD

**Discussion by rebeccamccabe** [sandialabs/WecOptTool#196] [Q&A]
Hi,
I am thinking about how to formulate my problem in WECOptTool - decisions of what objectives and constraints to give to the pseudospectral inner optimizer, given my complicated outer optimization loop. I can think of possible formulations that provide the PS more or less complicated / nonconvex objectives, and more or fewer constraints, which would in turn affect the convergence of my outer optimization loop. To weigh the different options, I am wondering if you can comment on how well you have found the PS optimization to perform for nonconvex objectives and with various constraints. Ie, does it consistently converge to the same local minimum with different initial guesses? What situations have you encountered where it fails to converge?

(I see I am starting the first Q&A discussion, so please let me know if this is the most appropriate place for this type of question). Thanks!

**Comment by rgcoe** [comment]
Thanks for your interest and question @rebeccamccabe. We have certainly observed some factors that affect convergence of pseudo-spectral problem, but I don't know that we've really explored this sufficiently to claim a complete understanding. If you want to pose some of the specific ideas you're considering we can certainly let you know what we think.

**Comment by rebeccamccabe** [comment]
Thanks! I am in the process of making up some block diagrams to more concretely show the possible structures I am thinking of, which I will post in a few days once they are finished. As an overview, I'm thinking about whether the pseudospectral optimizer should optimize average energy production (which equates to optimizing LCOE for a fixed design), or optimize the more complicated objective function that represents overall techno-economic viability (including power variation, etc so it might result in a different controller). 

If the latter, then there is also the question of whether I directly provide the PS optimizer with constraints on maximum force, travel, etc as is done in the examples (where the value of these limits would be decided by the outer optimizer based on the relationship between max force and cost), or if I leave these maximums unconstrained and instead let the PS optimizer decide what force limit is appropriate by including powertrain cost in its objective. The first case would have a smaller feasible design space and therefore less likely to get stuck in a local minimum that isn't globally optimal, and would facilitate intuition on interpreting the results. The second case would have a larger feasible design space and probably require multiple starts, but would not have to solve for the constraints which is computationally simpler.

**Reply by rgcoe** [reply]
Interesting stuff! I'm looking forward to seeing your block diagrams.

It sounds like you plan to formulate your "outer" optimization as a single objective problem. If so, I think another factor for consideration is that the PS optimization problem needs to be single objective (may be possible to change this, but would take some concerted work). Some of the structures you suggest above would allow you to use a multi-objective problem for the "outer" optimization, which I find somewhat attractive given the challenge/uncertainty distilling LCOE values from a device design that has a limited level of detail.

Not sure if you've already seen it, but Section 3.3 in this paper looks at something along the lines of what you're discussing, albeit with a more simple problem than you are considering: https://doi.org/10.1007/s40722-020-00181-9
